### ISSUE TITLE
Query: Short circuiting when left-over code is server correlated

### ISSUE THREAD

**Issue by smitpatel** [aspnet/EntityFrameworkCore#13859]
True port of #5736 for other expressions

This is different from #13068 in terms that #13068 talks about short-circuiting when the whole expression is client evaluatable. Which we solved by evaluating largest subtree possible. (hence compiler did short-circuit for us).

This issue is about expressions like this
```C#
List<Blog> values = null;
bool? test = false;
db.Blogs.Select(b => new { Test = (test ?? values.Select(e => e.Id).Contains(b.Id)) }).ToList();
```
Here we cannot evaluate `values.Select(e => e.Id)` since it throws NRE but we shouldn't need to evaluate it given test is there. We won't pick whole subtree because presence of `b.Id`

**Comment by fschlaef** [comment]
@smitpatel As of `3.0.0-preview5.19227.1` this is still broken.

This was originally punted for 3.0.0 but now it is in Backlog, is a fix still to be expected ?

Updated repro :

```csharp
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Debug;
using System;
using System.Linq;

namespace _13859Repro
{
    class Foo
    {
        public DateTime DateTime { get; set; }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var context = new MyContext())
            {
                //context.Database.EnsureDeleted();
                context.Database.EnsureCreated();
                context.Customer.Add(new Customer());
                context.SaveChanges();

                Foo foo = null;
                DateTime now = DateTime.UtcNow;

                // All of those are OK
                context.Customer.Select(c => new { Data = foo != null ? foo.DateTime : new DateTime(2018, 02, 02) }).FirstOrDefault();
                context.Customer.Select(c => new { Data = foo != null ? c.Date : DateTime.UtcNow }).FirstOrDefault();
                context.Customer.Select(c => new { Data = foo != null ? DateTime.UtcNow : c.Date }).FirstOrDefault();
                context.Customer.Select(c => new { Data = foo != null ? foo.DateTime : now }).FirstOrDefault();

                // But those are KO 
                context.Customer.Select(c => new { Data = foo != null ? foo.DateTime : DateTime.UtcNow }).FirstOrDefault();
                context.Customer.Select(c => new { Data = foo != null ? foo.DateTime : c.Date }).FirstOrDefault();
            }
        }
    }

    public class MyContext : DbContext
    {
        public virtual DbSet<Customer> Customer { get; set; }

        private static readonly LoggerFactory Logger = new LoggerFactory(new[] { new DebugLoggerProvider() });

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            string connectionString = "Server=.;Database=Repro13859;Trusted_Connection=True;MultipleActiveResultSets=true";

            optionsBuilder.UseSqlServer(connectionString)
                .EnableSensitiveDataLogging()
                .UseLoggerFactory(Logger);
        }
    }


    public partial class Customer
    {
        public int Id { get; set; }
        public DateTime Date { get; set; }
    }
}
```csharp

**Comment by ajcvickers** [comment]
@fschlaef We still plan to fix this, but the fix may come after the 3.0 release because we currently have too much work scheduled for 3.0 so we are unfortunately having to punt some issues.

**Comment by smitpatel** [comment]
TypeBinaryExpression can also short-circuit.
Something like this
```C#
IEntity entity = new Blog();
(e => entity is Post : ((Post)(e)).PostTitle : "")
```
